lib/UnexpectedExpressMocker.js

```diff
@@ -0,0 +1,366 @@
+const makePromise = require('unexpected/lib/makePromise');
+const http = require('http');
+const messy = require('messy');
+const stream = require('stream');
+const qs = require('qs');
+
+const classifyRequestBodyAndUpdateHttpRequest = require('./classifyRequestBodyAndUpdateHttpRequest');
+const errors = require('./errors');
+const resolveRequestBody = require('./resolveRequestBody');
+
+function applyRequestBodyToToReq(requestBodyState, req, httpRequest) {
+  if (requestBodyState) {
+    const { hasFormData, hasStream, requestBody } = requestBodyState;
+
+    if (hasStream) {
+      if (hasFormData) {
+        // form-data pauses its streams by default for some reason:
+        setImmediate(() => {
+          requestBody.resume();
+        });
+      }
+      const requestBodyChunks = [];
+      requestBody
+        .on('data', chunk => {
+          if (!Buffer.isBuffer(chunk)) {
+            chunk = Buffer.from(chunk, 'utf-8');
+          }
+          requestBodyChunks.push(chunk);
+          req.push(chunk);
+        })
+        .on('end', () => {
+          httpRequest.unchunkedBody = Buffer.concat(requestBodyChunks);
+          req.push(null);
+        })
+        .on('error', err => {
+          req.emit('error', err);
+        });
+    } else {
+      setImmediate(() => {
+        // To work around nodejs v0.10.x issue with old-style streams, see also https://github.com/stream-utils/raw-body/pull/34
+        req.push(requestBody);
+        req.push(null);
+      });
+    }
+  } else {
+    req.push(null);
+  }
+}
+
+function determineStatusProperty(obj) {
+  for (const prop of ['statusCode', 'status']) {
+    if (typeof obj[prop] === 'number') {
+      return prop;
+    }
+  }
+  return null;
+}
+
+class UnexpectedExpressMocker {
+  constructor(subject) {
+    if (!subject.handle || !subject.set) {
+      // This check is from the lib/application file in express @ 4.10.2.
+      // If we get inside here, we have something that is not an express app
+      // https://github.com/strongloop/express/blob/661435256384165bb656cb7b6046b4138ca24c9e/lib/application.js#L186
+      subject = require('express')().use(subject);
+    }
+
+    this.subject = subject;
+  }
+
+  mock(options) {
```

Comment:
Could `options` be destructured here?

```suggestion
  mock({ request, expectedErrorPassedToNext }) {
```

Reply:
I'm ok doing this, but it does me we cant throw a useful message if someone calls `.mock(null)` (which was the test I happened to just go and write while adding a check that a request description was supplied). What do you think?

Reply:
I think that's fine, it seems a bit weird for someone to call `.mock(null)` and expect to be treated nicely? ;)

That being said the destructuring is not hugely important to me, so I'm not going to insist ;)

Reply:
Ok cool - I've tried to make the mockers handle stuff like this, mostly for "niceness" if people pick up these tools (once separated). I can could see someone trying it to perhaps mean no request. I'll use destructuring but within the method and add a nice error message. That way it should remain consistent with its -mitm counterpart :)

Reply:
Ok - double checked and we've got tests that check that, in the absence of a specified requests, it'll default to a GET and we assert 404 is returned. As such, I will leave this as-is and we can tighten it as part of moving forward :)